include/seqan3/alignment/pairwise/policy/find_optimum_policy.hpp

```diff
@@ -0,0 +1,144 @@
+ * \tparam traits_type      A traits type that determines which cells should be considered for the optimum.
+ *                          Defaults to seqan3::detail::default_find_optimum_trait.
+ *
+ * \details
+ *
+ * This class determines the matrix wide optimum. The search space can be further refined using the
+ * `traits_type` which configures the search space of the alignment matrix.
+ */
+template <typename derived_t, typename traits_type = default_find_optimum_trait>
+class find_optimum_policy
+{
+
+private:
+
+    //!\brief Befriends the derived class to grant it access to the private members.
+    friend derived_t;
+
+    /*!\name Constructor, destructor and assignment
+     * \{
+     */
+    constexpr find_optimum_policy() = default;
+    constexpr find_optimum_policy(find_optimum_policy const &) = default;
+    constexpr find_optimum_policy(find_optimum_policy &&) = default;
+    constexpr find_optimum_policy & operator=(find_optimum_policy const &) = default;
+    constexpr find_optimum_policy & operator=(find_optimum_policy &&) = default;
+    ~find_optimum_policy() = default;
+    //!}
+
+protected:
+
+    /*!\brief Checks every cell of the dynamic programming matrix.
+     * \tparam score_t The type of the score.
+     * \param[in]     val       The current value.
+     * \param[in,out] optimum   The current optimum to compare with.
+     *
+     * \details
+     *
+     * This function resolves to a "NO-OP" function if the trait for searching every cell is set to std::false_type.
+     */
+    template <typename score_t>
+    constexpr void check_score([[maybe_unused]] score_t const val,
+                               [[maybe_unused]] score_t & optimum) const noexcept
+    {
+        if constexpr (traits_type::find_in_every_cell_type::value)
+            optimum = std::max(optimum, val);
+    }
+
+    /*!\brief Checks the last cell of the last row of the dynamic programming matrix.
+     * \tparam score_t The type of the score.
+     * \param[in]     val       The current value.
+     * \param[in,out] optimum   The current optimum to compare with.
+     *
+     * \details
+     *
+     * This function resolves to a "NO-OP" function if the trait for searching the last row is set to std::false_type.
+     * Due to a column based iteration layout this computes only the one cell at a time. The alignment algorithm
+     * takes care of calling this function for the appropriate cells.
+     */
+    template <typename score_t>
+    constexpr void check_score_last_row([[maybe_unused]] score_t const val,
+                                        [[maybe_unused]] score_t & optimum) const noexcept
+    {
+        if constexpr (traits_type::find_in_last_row_type::value)
+            optimum = std::max(optimum, val);
+    }
+
+    /*!\brief Checks the complete last column for the optimal score.
+     * \tparam rng_t   The type of the last column. Must be at least an input range; must model
```

Comment:
If it has to model `std::ranges::BidirectionalRange` it also models `std::ranges::InputRange`

Reply:
Yes. So it will always be `at least an input range` since it has to `model std::ranges::BidirectionalRange` ... so you don't have to write `at least an input range`.
Or am I missing something?

Reply:
Oh, I did not see what you are referring to. Yes the documentation needs to be adapted.